### PR TITLE
Adding SDK API for tab ready success and fail notification

### PR DESCRIPTION
**Pull request by sachingupta**
Adding SDK API for tab ready success and fail notification
Tab can notify to host if tab is ready to interact or failed with reason.

### CODE REVIEW

File: src/public/publicAPIs.ts [hunk 1]
```diff
@@ -3,92 +3,93 @@ import { GlobalVars } from "../internal/globalVars";
...
-    return;
-  }
-
-  GlobalVars.initializeCalled = true;
+export function initialize(hostWindow: any = window): AppLoadEvent {
+  // Independent components might not know whether the SDK is initialized so might call it to be safe.
+  // Just no-op if that happens to make it easier to use.
+  if (!GlobalVars.initializeCalled) {
+    GlobalVars.initializeCalled = true;
+
```
Comment (ydogandjiev) [inline src/public/publicAPIs.ts]: NIT: Delete extra empty line

File: src/public/publicAPIs.ts [hunk 2]
```diff
@@ -317,3 +318,13 @@ export function navigateToTab(tabInstance: TabInstance, onComplete?: (status: bo
   GlobalVars.callbacks[messageId] = onComplete ? onComplete : getGenericOnCompleteHandler(errorMessage);
 }
 
+export class AppLoadEvent implements IAppLoadEvent {
+  public notifySuccess(): void {
+    ensureInitialized();
+    sendMessageRequest(GlobalVars.parentWindow, "onTabShow.success", [version]);
```
Comment (ydogandjiev) [inline src/public/publicAPIs.ts]: NIT: I think it would be better if we called this `appLoadEvent.success` or `initialize.success`

Comment (ydogandjiev) [reply src/public/publicAPIs.ts]: Same for the failure case below

File: src/public/interfaces.ts [hunk 3]
```diff
@@ -394,3 +394,20 @@ export interface OpenConversationRequest {
...
+export interface IAppLoadEvent {
+  /**
+   * Indicates that the underlying resource has been created and the settings can be saved.
+   */
+  notifySuccess(): void;
+  /**
+   * Indicates that creation of the underlying resource failed and that the settings cannot be saved.
+   * @param reason Specifies a reason for the failure. If provided, this string is displayed to the user; otherwise a generic error is displayed.
+   */
+  notifyFailure(appLoadFailReason: IAppLoadFailReason): void;
```
Comment (ydogandjiev) [inline src/public/interfaces.ts]: Can we add an optional string parameter called something like `correlationTag` or `correlationId` that we can use to correlate telemetry between us and our partners? We can sanitize it for PII data and choose to only log this for 1P/2P apps to reduce the risk of PII data getting into our pipeline.

Comment (ydogandjiev) [reply src/public/interfaces.ts]: Oh, oops... I see that's what the `message` is for. That should be fine.

File: src/public/interfaces.ts [hunk 4]
```diff
@@ -394,3 +394,20 @@ export interface OpenConversationRequest {
   */
   onCloseConversation?: (subEntityId: string, conversationId?: string) => void;
 }
+
+export interface IAppLoadFailReason {
+  reason: AppLoadFailReasonTypes;
+  message?: string;
+}
+
+export interface IAppLoadEvent {
```
Comment (ydogandjiev) [inline src/public/interfaces.ts]: NIT: I think `IAppInitEvent` or `IAppInitializationEvent` might be a better name for this

File: src/public/constants.ts [hunk 5]
```diff
@@ -31,4 +31,10 @@ export const enum TaskModuleDimension {
   Large = "large",
   Medium = "medium",
   Small = "small"
+}
+
+export const enum AppLoadFailReasonTypes {
+  AuthFailed = "AuthFailed",
+  TimeOut = "TimeOut",
```
Comment (aosolis) [inline src/public/constants.ts]: "timeout" is a single word, so I think this should be cased "Timeout"

**Pull request merged**